teps/0011-redirecting-step-inputs-and-outputs.md

```diff
@@ -0,0 +1,143 @@
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+    - [Example Usage](#example-usage)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Consuming outputs of a step in another step is a common pattern in writing Tasks. However, this is currently tedious to do. Task authors have to overwrite the image entrypoint with either `sh -c` or the `script` field to wrap up the command to run with an explicit I/O redirection. This is not even possible if the image does not come with a shell, e.g., images built by [`ko`](https://github.com/google/ko).
+
+To achieve the functionality of I/O redirection between steps and even Tasks, Tekton will add new fields to steps for Task authors to specify paths to store stdout/stderr to and read stdin from.
+
+## Motivation
+
+Allow Task authors to support the following use cases:
+
+* Run image `gcr.io/k8s-staging-boskos/boskosctl` with args in a step and process its output through another `jq` image in a subsequent step to acquire the project name without overwriting the image entrypoint, building custom image with both utilities, or looking into container logs ([pipeline#2925](https://github.com/tektoncd/pipeline/issues/2925)).
+
+* Run image `alpine/git` with args `["describe", "--tags"]` in a step and extract the stdout as a Task result without overwriting the image entrypoint to be used by subsequent tasks.
+
+* In general, enable Task authors to access stdout/stderr of a step easily in another step in the same or different Task, without requiring authors to: 1) overwrite image entrypoints; 2) write wrapping scripts solely to do I/O redirection; and 3) build wrapper images to include all needed utilities.
+
+By enabling redirection between steps and Task results, it is much easier for Task authors to composite utility images in a Task.
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step through stdin.
+1. Allow a step to produce a Task result through its stdout or stderr.
+
+### Non-Goals
+
+1. Parse stdout of a step into a structured format (e.g., JSON) and extract information from certain fields.
+
+## Requirements
+
+* Add new fields for a step to specify a file to read from as stdin and files to write to as stdout/stderr.
+* Enhance the `entrypoint` binary to redirect stdin from and stdout/stderr to specified files.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Path to redirect stdin from on container's local filesystem.
+    // +optional
+    Stdin string `json:"stdin,omitempty"`
+    // Path to redirect stdout to on container's local filesystem.
+    // +optional
+    Stdout string `json:"stdout,omitempty"`
```

Comment:
I'd like to consider the alternative where the path is a convention instead of a configuration. This removes choice from the user, and doesn't increase the API surface.

Reply:
I'm not convinced yet if removing choice from user is good. The only main benefit of using "canonical paths" is that we will ensure that they can read them from another container, but that's not always needed. The major drawback is that we would lose users' "hints" on if they want to capture stdout/stderr. And if they're invoking some tooling that reads files from a specific path, they'll need an extra step to copy the stdout to the target path, which is doable but inconvenient.

Reply:
Adding this as another option to the proposal.